pkg/storage/s3.go

```diff
@@ -363,13 +363,20 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
 // The first argument is the file path that can be used in `Open`
 // function; the second argument is the size in byte of the file determined
 // by path.
-func (rs *S3Storage) WalkDir(ctx context.Context, dir string, listCount int64, fn func(string, int64) error) error {
+func (rs *S3Storage) WalkDir(ctx context.Context, fn func(string, int64) error) error {
 	var marker *string
-	prefix := rs.options.Prefix + dir
+	prefix := rs.options.Prefix
 	maxKeys := int64(1000)
-	if listCount > 0 {
+
+	dir, ok := ctx.Value("subDir").(string)
```

Comment:
I think it's better to put these parameters into one struct and provide a func to make the context so we can avoid set the string keys, something like:
```
type s3StoreCtxKey struct{}

type S3Context struct {
	dir string
	listCount int64
}

func WithS3Context(ctx context.Context, s S3Context) context.Context {
	return context.WithValue(ctx, s3StoreCtxKey{}, s)
}
```